Add block_in_if lint, #434

jmquigs opened the pull request:
For discussion, not quite ready yet...

Manishearth commented on src/block_in_if.rs:
```diff
@@ -0,0 +1,51 @@
+        if let ExprIf(ref check, ref then, _) = expr.node {
+            if let ExprBlock(ref block) = check.node {
+                if block.stmts.is_empty() {
+                    // TODO: its impossible to have an if block with no statements and no
```
Yes, it's impossible, since it  won't type check.

Manishearth commented on src/block_in_if.rs:
```diff
@@ -0,0 +1,51 @@
+                } else {
+                    // move block higher
+                    span_help_and_lint(cx, BLOCK_IN_IF, check.span,
+                        "move block out of condition",
```
text should be clearer

Manishearth commented on src/block_in_if.rs:
```diff
@@ -0,0 +1,51 @@
+
+                        span_help_and_lint(cx, BLOCK_IN_IF, check.span,
+                            "omit braces around single expression condition",
+                            &format!("try\nif {} {} ... ", snippet_block(cx, ex.span, ".."),
```
we should just use `snippet` instead of `snippet_block`, the latter adds braces sometimes

jmquigs replied:
I tried using snippet(), but it seems to preserve block indention, which looks funky when they get printed out.  I looked at using trim_multiline() to fix this, but then I realized that is exactly what snippet_block() does.  Actually, I'm not seeing how snippet_block can add braces, since it just appears to do snippet() + trim_multiline().

Manishearth replied:
Ah, okay then.

Manishearth commented:
looks good overall

Manishearth commented:
Tests don't seem to pass. Also, run `update_lints.py` once you're done and commit the changes

Manishearth commented on src/block_in_if.rs:
```diff
@@ -0,0 +1,51 @@
+
+// TODO: maybe use two lints for this?
+declare_lint! {
+    pub BLOCK_IN_IF, Warn,
```
block_in_if_condition, maybe?

llogiq commented:
No, I meant something like `if predicate(|x| { .. }) { ..}`.

jmquigs commented:
@llogiq, ok I will try to add a check for this case.  Btw I renamed the lint and split it into two lints, I now have the record for longest named lint in the code :stuck_out_tongue_closed_eyes:

jmquigs commented:
I implemented the check for closures.  It does indeed catch cases like this:
```rust
if v == 3 && sky == "blue" && predicate(|x| { let target = 3; x == target }, v) {}
```
I'd like to look for some false positives though, can you guys recommend any codebases that are good to check?

jmquigs commented:
Looks like a nightly update has rocked clippy's world

Manishearth commented:
Manually merged as https://github.com/Manishearth/rust-clippy/commit/d4cf288b385b9c58cfa0d4c97b3b548bc21b4923

Thanks! :smile:
